template/src/schemas/schema.ts.js

```diff
@@ -0,0 +1,41 @@
+import { File } from '@asyncapi/generator-react-sdk';
+const {TypeScriptGenerator, FormatHelpers} = require('@asyncapi/generator-model-sdk');
+
+/**
+ * @typedef RenderArgument
+ * @type {object}
+ * @property {AsyncAPIDocument} asyncapi received from the generator.
+ */
+
+/**
+ * Custom model preset to ensure property names are lower cased
+ */
+const preset = {
+  interface: {
+    async property({ renderer, model, propertyName, property }) {
+      const signature = renderer.renderTypeSignature(property, { isRequired: model.isRequired(propertyName) });
+      return `${propertyName}${signature};`;
+    }
+  }
+};
+
+/**
+ * Render all schema models
+ * @param {RenderArgument} param0 
+ * @returns 
+ */
+export default async function schemaRender({ asyncapi }) {
+  const typescriptGenerator = new TypeScriptGenerator({modelType: 'interface', presets: [preset]});
+  const generatedModels = await typescriptGenerator.generate(asyncapi);

```

Comment:
Why do you generate (and then iterate) all models from here? Why not using something like `$$objectSchema$$.js` and then let the Generator do the magic for you?

Reply:
The `$$objectSchema$$ is just a temporary solution, any advanced validation rules will never work with that file template.

The model generator library infer the definition of models based on validation rules and decide when separate models need to be generated. The generator just uses object validation rules directly which will never work correctly.

Reply:
Ok so why not using `$$schema$$` and then just rending a model for a single schema? In the code, you can check if the schema actually qualifies to be generated or not.

The reason I'm insisting is that —to me— seems non-sense to iterate over all the schemas of the asyncapi file, again. The Generator does that for you, take advantage of it.

Reply:
@fmvilas `$$schema$$` only iterate component schemas and not every message schema. 

> The reason I'm insisting is that —to me— seems non-sense to iterate over all the schemas of the asyncapi file, again

I am not though, I just feed the AsyncAPI document to the model generator library and let it handle which models should be generated. I then iterate those models that it want to generate and create a file for each.

The generator file templates are not built for generating models, if anything it would maybe be an idea with a `$$model$$` file template. But it all comes down to which integration we find most fitting, and I am not sure this would ever be used, especially when template developers need control over the generation process 😄

Reply:
> `$$schema$$` only iterate component schemas and not every schema in use.

Ok, use `$$everySchema$$` instead.

> Also to me using any file template is non-sense to me, especially since all it gives me are schemas. I would need the same functionality anyway, and do the exact same thing I am doing now regardless. Instead of providing an AsyncAPI document, I would provide the JSON schema instead.

You don't have to do the same. You get a schema and the asyncapi file (for context). The only thing you'd have to do is to render a single schema (the one that is passed). Then, all the schemas will be generated because the Generator will call your renderer once per schema definition being used: https://github.com/asyncapi/generator/blob/master/lib/generator.js#L515. And that's my whole point, you don't need to iterate over the schemas **once again**. The Generator is doing this for you. Use it.

> The generator file templates are not built for generating models

That was literally the purpose it was created for 😂 Then other filename templates appeared afterward.

Reply:
> Ok, use $$everySchema$$ instead.

Reply:
Let me give you an example on why I am reluctant to change to file templates. Given a message payload defined as such: 
```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "test1"
  "type": "object",
  "properties": {
    "street_address": {
      "type": "object",
      "$id": "test2"
      "properties": {
        "floor": {
          "$id": "test3"
          "type": "number"
        }
      }
    },
    "country": {
       "$id": "test4"
      "enum": ["United States of America", "Canada"]
    }
  }
}
```

* With `$$everySchema$$` the generator will try to generate all 4 schemas. However by schema `test1` the model generator will already have generated all the other schemas and render them useless.
* With `$$schema$$` the generator will not generate anything since nothing is defined under components.
* with `$$objectSchema$$` the generator will try to generate `test2` and `test1`. However by schema `test1` the model generator will already have generated both models render the second iteration useless.

What is possible, is to do it with `$$message$$` or `$$channel$$` but again the code would remain somewhat the same with the only difference that I would pass the JSON Schema payload the model generator instead of the entire AsyncAPI document. To me this seems more cumbersome and confusing then the current code, but I might be biased 🤷

Reply:
@magicmatatjahu what are your thoughts, would you suggest I switch to filetemplate? 🤔

Reply:
For me, the `for loop` is better 😄 There is no need to check whether a given model has already been generated - sometimes the given model may not be generated (examples were given by Jonas). For me, iterating and returning an array of files is clearer.

Reply:
>   * With $$everySchema$$ the generator will try to generate all 4 schemas. However, by schema test1 the model generator will already have generated all the schemas.
> * With $$schema$$ the generator will not generate anything since nothing is defined under components.
> * with $$objectSchema$$ the generator will try to generate test2 and test1. However by schema test1 the model generator will already have generated both models render the second iteration useless.

Ok, I see. Then just keep it this way 👍